File: tk_advance_vehicle_repair/models/advance_vehicle_repair_dashboard.py
# -*- coding: utf-8 -*-
# Copyright 2022-Today TechKhedut.
# Part of TechKhedut. See LICENSE file for full copyright and licensing details.
from odoo import models, fields, api, _


class AdvanceVehicleRepairDashboard(models.Model):
    """Advance Vehicle Repair Dashboard"""
    _name = "advance.vehicle.repair.dashboard"
    _description = __doc__

    @api.model
    def get_advance_vehicle_repair_dashboard(self):
        total_vehicle_booking = self.env['vehicle.booking'].sudo().search_count([])
        vehicle_inspection = self.env['vehicle.booking'].sudo().search_count(
            [('booking_stages', '=', 'vehicle_inspection')])
        vehicle_repair = self.env['vehicle.booking'].sudo().search_count([('booking_stages', '=', 'vehicle_repair')])
        vehicle_inspection_repair = self.env['vehicle.booking'].sudo().search_count(
            [('booking_stages', '=', 'vehicle_inspection_repair')])
        booking_cancel = self.env['vehicle.booking'].sudo().search_count([('booking_stages', '=', 'cancel')])

        total_inspection_job_card = self.env['inspection.job.card'].sudo().search_count([])
        inspection_in_progress = self.env['inspection.job.card'].sudo().search_count([('stages', '=', 'b_in_progress')])
        inspection_complete = self.env['inspection.job.card'].sudo().search_count([('stages', '=', 'c_complete')])
        inspection_cancel = self.env['inspection.job.card'].sudo().search_count([('stages', '=', 'd_cancel')])

        repair_job_card = self.env['repair.job.card'].sudo().search_count([])
        assign_to_technician = self.env['repair.job.card'].sudo().search_count(
            [('stages', '=', 'assign_to_technician')])
        in_diagnosis = self.env['repair.job.card'].sudo().search_count([('stages', '=', 'in_diagnosis')])
        supervisor_inspection = self.env['repair.job.card'].sudo().search_count(
            [('stages', '=', 'supervisor_inspection')])
        hold = self.env['repair.job.card'].sudo().search_count([('stages', '=', 'hold')])
        complete = self.env['repair.job.card'].sudo().search_count([('stages', '=', 'complete')])
        cancel = self.env['repair.job.card'].sudo().search_count([('stages', '=', 'cancel')])

        booking_source_direct = self.env['vehicle.booking'].sudo().search_count([('booking_source', '=', 'direct')])
        booking_source_website = self.env['vehicle.booking'].sudo().search_count([('booking_source', '=', 'website')])

        inspection_job_card_full_inspection = self.env['inspection.job.card'].sudo().search_count(
            [('inspection_type', '=', 'full_inspection')])
        inspection_job_card_specific_inspection = self.env['inspection.job.card'].sudo().search_count(
            [('inspection_type', '=', 'specific_inspection')])

        repair_job_card_details = [
            ['Job Cards', 'Assign to Technician', 'In Diagnosis', 'Supervisor Inspection', 'Hold', 'Completed',
             'Cancelled'],
            [repair_job_card, assign_to_technician, in_diagnosis, supervisor_inspection, hold, complete, cancel]]

        booking_details = [['Vehicle Inspection', 'Vehicle Repair', 'Vehicle Inspection and Repair'],
                           [vehicle_inspection, vehicle_repair, vehicle_inspection_repair]]

        booking_source = [['Direct', 'Website'], [booking_source_direct, booking_source_website]]

        inspection_job_card_details = [['Full Inspection', 'Specific Inspection'],
                                       [inspection_job_card_full_inspection, inspection_job_card_specific_inspection]]

        data = {
            'total_vehicle_booking': total_vehicle_booking,
            'vehicle_inspection': vehicle_inspection,
            'vehicle_repair': vehicle_repair,
            'vehicle_inspection_repair': vehicle_inspection_repair,
            'booking_cancel': booking_cancel,
            'total_inspection_job_card': total_inspection_job_card,
            'inspection_in_progress': inspection_in_progress,
            'inspection_complete': inspection_complete,
            'inspection_cancel': inspection_cancel,
            'repair_job_card_details': repair_job_card_details,
            'booking_details': booking_details,
            'booking_source': booking_source,
            'inspection_job_card_details': inspection_job_card_details,
            'common_fuel_used_in_vehicle': self.get_most_common_fuels(),
        }
        return data

    def get_most_common_fuels(self):
        top_five_fuel = {}
        for group in self.env['vehicle.booking'].read_group([], ['vehicle_fuel_type_id'], ['vehicle_fuel_type_id'],
                                                            orderby="vehicle_fuel_type_id DESC"):
            fuel_type = self.env['vehicle.fuel.type'].sudo().browse(int(group['vehicle_fuel_type_id'][0])).name
            top_five_fuel[fuel_type] = group['vehicle_fuel_type_id_count']
        top_five_fuel = dict(sorted(top_five_fuel.items(), key=lambda item: item[1], reverse=True))
        return [list(top_five_fuel.keys()), list(top_five_fuel.values())]

File: tk_advance_vehicle_repair/static/src/js/advance_vehicle_repair_dashboard.js
/** @odoo-module **/
import {registry} from "@web/core/registry";
import {Layout} from "@web/search/layout";
import {getDefaultConfig} from "@web/views/view";
import {useService} from "@web/core/utils/hooks";
import { useDebounced } from "@web/core/utils/timing";
import { session } from "@web/session";
import {Domain} from "@web/core/domain";
import {sprintf} from "@web/core/utils/strings";

const {Component, useSubEnv, useState, onMounted, onWillStart, useRef} = owl;
import {loadJS, loadCSS} from "@web/core/assets"

class AdvanceVehicleDashboard extends Component {
    setup() {
        this.rpc = useService("rpc");
        this.action = useService("action");
        this.orm = useService("orm");

        this.state = useState({
            bookingStats: {'total_vehicle_booking': 0, 'vehicle_inspection': 0, 'vehicle_repair': 0, 'vehicle_inspection_repair': 0, 'booking_cancel': 0},
            inspectionJobCardStatus : {'total_inspection_job_card': 0, 'inspection_in_progress': 0, 'inspection_complete': 0, 'inspection_cancel': 0},
            repairJobCardStatus: {'x-axis': [], 'y-axis': []},
            bookingDetailsStatus: {'x-axis': [], 'y-axis': []},
            bookingSource: {'x-axis': [], 'y-axis': []},
            inspectionJobCardDetails: {'x-axis': [], 'y-axis': []},
            commonVehicleFuels: {'x-axis': [], 'y-axis': []},
        });

        useSubEnv({
            config: {
                ...getDefaultConfig(),
                ...this.env.config,
            },
        });

        this.repairJobCardStatus = useRef('vehicle_repair_job_card_details');
        this.bookingDetailsStatus = useRef('booking_details_graph');
        this.bookingSource = useRef('booking_source_graph');
        this.inspectionJobCardDetails = useRef('job_card_inspection_type');
        this.commonVehicleFuels = useRef('most_common_fuel_used_in_vehicle');

        onWillStart( async () => {
            await loadJS('/tk_advance_vehicle_repair/static/src/js/apexcharts.js');
            let bookingData = await this.orm.call('advance.vehicle.repair.dashboard', 'get_advance_vehicle_repair_dashboard', []);
            if(bookingData){
                this.state.bookingStats = bookingData;
                this.state.inspectionJobCardStatus = bookingData;
                this.state.repairJobCardStatus = {'x-axis': bookingData['repair_job_card_details'][0], 'y-axis': bookingData['repair_job_card_details'][1]}
                this.state.bookingDetailsStatus = {'x-axis': bookingData['booking_details'][0], 'y-axis': bookingData['booking_details'][1]}
                this.state.bookingSource = {'x-axis': bookingData['booking_source'][0], 'y-axis': bookingData['booking_source'][1]}
                this.state.inspectionJobCardDetails = {'x-axis': bookingData['inspection_job_card_details'][0], 'y-axis': bookingData['inspection_job_card_details'][1]}
                this.state.commonVehicleFuels = {'x-axis': bookingData['common_fuel_used_in_vehicle'][0], 'y-axis': bookingData['common_fuel_used_in_vehicle'][1]}
            }
        });
        onMounted(() => {
            this.renderRepairJobCardStatusGraph();
            this.renderBookingDetailsGraph();
            this.renderBookingSource();
            this.renderInspectionJobCardDetails();
            this.renderCommonVehicleFuels();
        })
    }

    renderRepairJobCardStatusGraph(){
        const options = {
            series: [
                {
                name: 'Status',
                data: this.state.repairJobCardStatus['y-axis'],
                }
            ],
            chart: {
                height: 400,
                type: 'bar',
            },
            colors: ['#f29e4c', '#f1c453', '#efea5a',  '#b9e769',  '#83e377', '#16db93', '#0db39e', '#048ba8', '#2c699a', '#54478c'],
            plotOptions: {
                bar: {
                    columnWidth: '10%',
                    distributed: true,
                }
            },
            dataLabels: {
                enabled: false
            },
            legend: {
                show: false
            },
            xaxis: {
                categories: this.state.repairJobCardStatus['x-axis'],
                labels: {
                    style: {
                        colors: ['#000000'],
                        fontSize: '13px'}
                }
            }
        };
        this.renderGraph(this.repairJobCardStatus.el, options);
    }

    renderBookingDetailsGraph() {
        const options = {
            series: this.state.bookingDetailsStatus['y-axis'],
            chart: {
                type: 'pie',
                height: 410
            },
            dataLabels: {
                enabled: false
            },
            labels: this.state.bookingDetailsStatus['x-axis'],
            legend: {
                position: 'bottom',
            },
        };
        this.renderGraph(this.bookingDetailsStatus.el, options);
    }

    renderBookingSource(){
        const options = {
            series: [{
                name: 'Status',
                data: this.state.bookingSource['y-axis'],
            }],
            chart: {
                height: 400,
                type: 'bar',
            },
            colors: ['#2c699a', '#54478c'],
            plotOptions: {
                bar: {
                    columnWidth: '10%',
                    distributed: true,
                }
            },
            dataLabels: {
                enabled: false
            },
            legend: {
                show: false
            },
            xaxis: {
                categories: this.state.bookingSource['x-axis'],
                labels: {
                    style: {
                        colors: ['#000000'],
                        fontSize: '13px'}
                }
            }
        };
        this.renderGraph(this.bookingSource.el, options);
    }

    renderInspectionJobCardDetails() {
        const options = {
            series: [{
                name: 'Status',
                data: this.state.inspectionJobCardDetails['y-axis'],
            }],
            chart: {
                height: 400,
                type: 'bar',
            },
            colors: ['#16db93', '#83e377', '#b9e769', '#efea5a', '#f1c453', '#f29e4c'],
            plotOptions: {
                bar: {
                    columnWidth: '10%',
                    distributed: true,
                }
            },
            dataLabels: {
                enabled: false
            },
            legend: {
                show: false
            },
            xaxis: {
                categories: this.state.inspectionJobCardDetails['x-axis'],
                labels: {
                    style: {
                        colors: ['#000000'],
                        fontSize: '13px'}
                }
            }
        };
        this.renderGraph(this.inspectionJobCardDetails.el, options);
    }

    renderCommonVehicleFuels(){
        const options = {
            series: this.state.commonVehicleFuels['y-axis'],
            chart: {
                width: 500,
                height: 500,
                type: 'polarArea',
            },
            yaxis: {
                labels: {
                    formatter: function (val) {
                    return val.toFixed(0);
                    }
                },
            },
            dataLabels: {
                enabled:true,
            },
            labels: this.state.commonVehicleFuels['x-axis'],
            stroke: {
                colors: ['#fff']
            },
            fill: {
                opacity: 0.8
            },
            responsive: [{
                breakpoint: 480,
                options: {
                    chart: {
                        width: 200
                    },
                    legend: {
                        position: 'bottom'
                    }
                }
            }]
        };
        this.renderGraph(this.commonVehicleFuels.el, options);
       }

    renderGraph(el, options){
        const graphData = new ApexCharts(el, options);
        graphData.render();
    }

    viewVehicleBooking(type){
        let domain,context;
        let name = this.getBookingStatus(type);
        if (type === 'all'){
            domain = []
        }else {
            domain = [['booking_stages', '=', type]]
        }
        context = { 'create':false }
        this.action.doAction({
            type: 'ir.actions.act_window',
            name: name,
            res_model: 'vehicle.booking',
            view_mode: 'kanban',
            views: [[false, 'kanban'], [false, 'list'], [false, 'form'], [false, 'pivot'], [false, 'activity']],
            target: 'current',
            context: context,
            domain: domain,
        });
    }

    getBookingStatus(type){
        let name;
        if(type === 'all') {
            name = 'Bookings'
        }else if(type === 'vehicle_inspection') {
            name = 'In Inspections'
        }else if(type === 'vehicle_repair') {
            name = 'In Repairs'
        } else if(type === 'vehicle_inspection_repair') {
            name = 'In Inspections and Repairs'
        }else if(type === 'cancel') {
            name = 'Booking Cancelled'
        }
        return name;
    }

    viewInspectionJobCard(type){
        let domain,context;
        let name = this.getInspectionJobCardStatus(type);
        if (type === 'all'){
            domain = []
        }else {
            domain = [['stages', '=', type]]
        }
        context = { 'create':false }
        this.action.doAction({
            type: 'ir.actions.act_window',
            name: name,
            res_model: 'inspection.job.card',
            view_mode: 'kanban',
            views: [[false, 'kanban'], [false, 'list'], [false, 'form'], [false, 'pivot'], [false, 'activity']],
            target: 'current',
            context: context,
            domain: domain,
        });
    }

    getInspectionJobCardStatus(type){
        let name;
        if(type === 'all') {
            name = 'Job Cards'
        }else if(type === 'b_in_progress') {
            name = 'Inspection In Progress'
        }else if(type === 'c_complete') {
            name = 'Inspection Completed'
        }else if(type === 'd_cancel') {
            name = 'Inspection Cancelled'
        }
        return name;
    }

}
AdvanceVehicleDashboard.template = "tk_advance_vehicle_repair.adv_dashboard";
registry.category("actions").add("advance_vehicle_repair_dashboard", AdvanceVehicleDashboard);